diff --git a/include/wordlist.h b/include/wordlist.h
index d4b722ec2..2af74f7ed 100644
--- a/include/wordlist.h
+++ b/include/wordlist.h
@@ -11,6 +11,8 @@
 
 size_t convert_from_hex (hashcat_ctx_t *hashcat_ctx, char *line_buf, const size_t line_len);
 
+int comb_word_transform (hashcat_ctx_t *hashcat_ctx, u8 *buf, int len, const bool iconv_enabled, const iconv_t iconv_ctx, char *iconv_tmp);
+
 void pw_pre_add       (hc_device_param_t *device_param, const u8 *pw_buf, const int pw_len, const u8 *base_buf, const int base_len, const int rule_idx);
 void pw_base_add      (hc_device_param_t *device_param, pw_pre_t *pw_pre);
 void pw_add_zerocopy  (hc_device_param_t *device_param, u8 *out_buf, const int pw_len);
diff --git a/src/backend.c b/src/backend.c
index 06b3bdb0f..259219207 100644
--- a/src/backend.c
+++ b/src/backend.c
@@ -4377,70 +4377,33 @@ int run_cracker (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_param, co
 
                 size_t line_len = fgetl (combs_fp, line_buf, HCBUFSIZ_LARGE);
 
-                line_len = convert_from_hex (hashcat_ctx, line_buf, line_len);
+                const int comb_len = comb_word_transform (hashcat_ctx, (u8 *) line_buf, (int) line_len, iconv_enabled, iconv_ctx, iconv_tmp);
 
-                if (line_len > PW_MAX) continue;
-
-                char *line_buf_new = line_buf;
-
-                char rule_buf_out[RP_PASSWORD_SIZE];
-
-                if (run_rule_engine (user_options_extra->rule_len_r, user_options->rule_buf_r))
+                if (comb_len < 0)
                 {
-                  if (line_len >= RP_PASSWORD_SIZE) continue;
-
-                  memset (rule_buf_out, 0, sizeof (rule_buf_out));
-
-                  const int rule_len_out = _old_apply_rule (user_options->rule_buf_r, user_options_extra->rule_len_r, line_buf, (u32) line_len, rule_buf_out);
-
-                  if (rule_len_out < 0)
+                  if (user_options->attack_mode == ATTACK_MODE_ASSOCIATION)
                   {
-                    if (user_options->attack_mode == ATTACK_MODE_ASSOCIATION)
+                    for (u32 association_salt_pos = 0; association_salt_pos < pws_cnt; association_salt_pos++)
                     {
-                      for (u32 association_salt_pos = 0; association_salt_pos < pws_cnt; association_salt_pos++)
-                      {
-                        status_ctx->words_progress_rejected[association_salt_pos] += 1;
-                      }
+                      status_ctx->words_progress_rejected[association_salt_pos] += 1;
                     }
-                    else
-                    {
-                      status_ctx->words_progress_rejected[salt_pos] += pws_cnt;
-                    }
-
-                    continue;
+                  }
+                  else
+                  {
+                    status_ctx->words_progress_rejected[salt_pos] += pws_cnt;
                   }
 
-                  line_len = rule_len_out;
-
-                  line_buf_new = rule_buf_out;
-                }
-
-                // do the on-the-fly encoding
-
-                if (iconv_enabled == true)
-                {
-                  char  *iconv_ptr = iconv_tmp;
-                  size_t iconv_sz  = HCBUFSIZ_TINY;
-
-                  if (iconv (iconv_ctx, &line_buf_new, &line_len, &iconv_ptr, &iconv_sz) == (size_t) -1) continue;
-
-                  line_buf_new = iconv_tmp;
-                  line_len     = HCBUFSIZ_TINY - iconv_sz;
+                  continue;
                 }
 
-                line_len = MIN (line_len, PW_MAX);
+                line_len = (size_t) comb_len;
 
                 u8 *ptr = (u8 *) device_param->combs_buf[i].i;
 
-                memcpy (ptr, line_buf_new, line_len);
+                memcpy (ptr, line_buf, line_len);
 
                 memset (ptr + line_len, 0, PW_MAX - line_len);
 
-                if (hashconfig->opts_type & OPTS_TYPE_PT_UPPER)
-                {
-                  uppercase (ptr, line_len);
-                }
-
                 if (combinator_ctx->combs_mode == COMBINATOR_MODE_BASE_LEFT)
                 {
                   if (hashconfig->opts_type & OPTS_TYPE_PT_ADD80)
@@ -4568,70 +4531,33 @@ int run_cracker (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_param, co
 
                 size_t line_len = fgetl (combs_fp, line_buf, HCBUFSIZ_LARGE);
 
-                line_len = convert_from_hex (hashcat_ctx, line_buf, line_len);
+                const int comb_len = comb_word_transform (hashcat_ctx, (u8 *) line_buf, (int) line_len, iconv_enabled, iconv_ctx, iconv_tmp);
 
-                if (line_len > PW_MAX) continue;
-
-                char *line_buf_new = line_buf;
-
-                char rule_buf_out[RP_PASSWORD_SIZE];
-
-                if (run_rule_engine (user_options_extra->rule_len_r, user_options->rule_buf_r))
+                if (comb_len < 0)
                 {
-                  if (line_len >= RP_PASSWORD_SIZE) continue;
-
-                  memset (rule_buf_out, 0, sizeof (rule_buf_out));
-
-                  const int rule_len_out = _old_apply_rule (user_options->rule_buf_r, user_options_extra->rule_len_r, line_buf, (u32) line_len, rule_buf_out);
-
-                  if (rule_len_out < 0)
+                  if (user_options->attack_mode == ATTACK_MODE_ASSOCIATION)
                   {
-                    if (user_options->attack_mode == ATTACK_MODE_ASSOCIATION)
+                    for (u32 association_salt_pos = 0; association_salt_pos < pws_cnt; association_salt_pos++)
                     {
-                      for (u32 association_salt_pos = 0; association_salt_pos < pws_cnt; association_salt_pos++)
-                      {
-                        status_ctx->words_progress_rejected[association_salt_pos] += 1;
-                      }
+                      status_ctx->words_progress_rejected[association_salt_pos] += 1;
                     }
-                    else
-                    {
-                      status_ctx->words_progress_rejected[salt_pos] += pws_cnt;
-                    }
-
-                    continue;
+                  }
+                  else
+                  {
+                    status_ctx->words_progress_rejected[salt_pos] += pws_cnt;
                   }
 
-                  line_len = rule_len_out;
-
-                  line_buf_new = rule_buf_out;
-                }
-
-                // do the on-the-fly encoding
-
-                if (iconv_enabled == true)
-                {
-                  char  *iconv_ptr = iconv_tmp;
-                  size_t iconv_sz  = HCBUFSIZ_TINY;
-
-                  if (iconv (iconv_ctx, &line_buf_new, &line_len, &iconv_ptr, &iconv_sz) == (size_t) -1) continue;
-
-                  line_buf_new = iconv_tmp;
-                  line_len     = HCBUFSIZ_TINY - iconv_sz;
+                  continue;
                 }
 
-                line_len = MIN (line_len, PW_MAX);
+                line_len = (size_t) comb_len;
 
                 u8 *ptr = (u8 *) device_param->combs_buf[i].i;
 
-                memcpy (ptr, line_buf_new, line_len);
+                memcpy (ptr, line_buf, line_len);
 
                 memset (ptr + line_len, 0, PW_MAX - line_len);
 
-                if (hashconfig->opts_type & OPTS_TYPE_PT_UPPER)
-                {
-                  uppercase (ptr, line_len);
-                }
-
                 /*
                 if (combinator_ctx->combs_mode == COMBINATOR_MODE_BASE_LEFT)
                 {
diff --git a/src/selftest.c b/src/selftest.c
index 7aba9ad99..2fc17d6fc 100644
--- a/src/selftest.c
+++ b/src/selftest.c
@@ -101,6 +101,11 @@ static int selftest_init (hashcat_ctx_t *hashcat_ctx, hc_device_param_t *device_
 
     pw.pw_len = (u32) pw_len;
 
+    if (hashconfig->opts_type & OPTS_TYPE_PT_UPPER)
+    {
+      uppercase ((u8 *) pw_ptr, pw.pw_len);
+    }
+
     if (device_param->is_cuda == true)
     {
       if (hc_cuMemcpyHtoD (hashcat_ctx, device_param->cuda_d_pws_buf, &pw, 1 * sizeof (pw_t)) == -1) return -1;
diff --git a/src/slow_candidates.c b/src/slow_candidates.c
index e004ce1fe..c38a32df1 100644
--- a/src/slow_candidates.c
+++ b/src/slow_candidates.c
@@ -291,27 +291,13 @@ void slow_candidates_next (hashcat_ctx_t *hashcat_ctx, void *extra_info)
 
     while (true)
     {
-      line_len = (u32) fgetl (combs_fp, line_buf, HCBUFSIZ_LARGE);
+      const int raw_len = (int) fgetl (combs_fp, line_buf, HCBUFSIZ_LARGE);
 
-      line_len = convert_from_hex (hashcat_ctx, line_buf, line_len);
+      const int comb_len = comb_word_transform (hashcat_ctx, (u8 *) line_buf, raw_len, hashcat_ctx->wl_data->iconv_enabled, hashcat_ctx->wl_data->iconv_ctx, hashcat_ctx->wl_data->iconv_tmp);
 
-      // post-process rule engine
+      if (comb_len < 0) continue;
 
-      if (run_rule_engine ((int) user_options_extra->rule_len_r, user_options->rule_buf_r))
-      {
-        if (line_len >= RP_PASSWORD_SIZE) continue;
-
-        char rule_buf_out[RP_PASSWORD_SIZE];
-
-        memset (rule_buf_out, 0, sizeof (rule_buf_out));
-
-        const int rule_len_out = _old_apply_rule (user_options->rule_buf_r, (int) user_options_extra->rule_len_r, line_buf, (int) line_len, rule_buf_out);
-
-        if (rule_len_out < 0) continue;
-
-        line_buf = rule_buf_out;
-        line_len = (u32) rule_len_out;
-      }
+      line_len = (u32) comb_len;
 
       break;
     }
diff --git a/src/wordlist.c b/src/wordlist.c
index 6ebfe3b19..170acb9a7 100644
--- a/src/wordlist.c
+++ b/src/wordlist.c
@@ -275,6 +275,56 @@ void get_next_word_std (char *buf, u64 sz, u64 *len, u64 *off)
   *len = sz;
 }
 
+int comb_word_transform (hashcat_ctx_t *hashcat_ctx, u8 *buf, int len, const bool iconv_enabled, const iconv_t iconv_ctx, char *iconv_tmp)
+{
+
+  const hashconfig_t         *hashconfig         = hashcat_ctx->hashconfig;
+  const user_options_t       *user_options       = hashcat_ctx->user_options;
+  const user_options_extra_t *user_options_extra = hashcat_ctx->user_options_extra;
+
+  len = (int) convert_from_hex (hashcat_ctx, (char *) buf, (size_t) len);
+
+  if (len > PW_MAX) return -1;
+
+  if (run_rule_engine ((int) user_options_extra->rule_len_r, user_options->rule_buf_r))
+  {
+    if (len >= RP_PASSWORD_SIZE) return -1;
+
+    char rule_buf_out[RP_PASSWORD_SIZE];
+
+    memset (rule_buf_out, 0, sizeof (rule_buf_out));
+
+    const int rule_len_out = _old_apply_rule (user_options->rule_buf_r, (int) user_options_extra->rule_len_r, (char *) buf, len, rule_buf_out);
+
+    if (rule_len_out < 0) return -1;
+
+    len = rule_len_out;
+
+    memcpy (buf, rule_buf_out, (size_t) len);
+  }
+
+  if (iconv_enabled == true)
+  {
+    char  *iconv_ptr = iconv_tmp;
+    size_t iconv_sz  = HCBUFSIZ_TINY;
+    char  *in_ptr    = (char *) buf;
+    size_t in_len    = (size_t) len;
+
+    if (iconv (iconv_ctx, &in_ptr, &in_len, &iconv_ptr, &iconv_sz) == (size_t) -1) return -1;
+
+    len = (int) (HCBUFSIZ_TINY - iconv_sz);
+
+    memcpy (buf, iconv_tmp, (size_t) len);
+  }
+
+  if (hashconfig->opts_type & OPTS_TYPE_PT_UPPER)
+  {
+    uppercase (buf, (size_t) len);
+  }
+
+  return MIN (len, PW_MAX);
+}
+
 void get_next_word (hashcat_ctx_t *hashcat_ctx, HCFILE *fp, char **out_buf, u32 *out_len)
 {
   user_options_t       *user_options       = hashcat_ctx->user_options;
